### ISSUE TITLE
Wrong way to split the dataset

### ISSUE THREAD

**Issue by darsvador** [rickyzhang82/V2Ray-Deep-Packet-Inspection#1]
An important result of this repository is that `V2Ray traffic data` can be easily detected by a CNN model. But I found that the method of dividing the data set seems to be wrong. The following code is in [this file](https://github.com/rickyzhang82/V2Ray-Deep-Packet-Inspection/blob/master/V2Ray%20Deep%20Packet%20Inspection.ipynb). The following code undersamples both the training set and the validation set. It's ok to undersample on the training set, but undersampling on the validation/test set is not such a good choice. This method of dividing the data set allows the classifier to cheat. Because the distribution of positive and negative examples of the test set has been artificially adjusted. For machine learning, what is important is the independence of the training set and the test set.

````


import numpy as np
import math
import os.path
from pathlib import Path
import glob
from tensorflow.keras.utils import Sequence

FIXED_PACKET_SIZE = 1500
NUM_OF_PACKETS_PER_FILE = 16
RESCALE_FACTOR = 1./255

# v2ray traffic tag
TRAINING_DATA_PERCENTAGE = 0.8
PACKET_FILE_EXT = '*.bin'


def rglob(data_root, file_ext):
    files = list()
    for filePath in Path(data_root).rglob(file_ext):
        files.append(str(filePath))
    return files


def binary_classification(packet_path, match_string=V2RAY_HOST_TAG):
    """Binary network traffic classification function

    :param packet_path: file path to packet
    :param match_string:
    :return: 1, if it is v2ray traffice. 0, otherwise.
    """
    if packet_path.find(match_string) != -1:
        return 1
    else:
        return 0


def generate_train_validation_packet_path_list(data_root, training_pct=TRAINING_DATA_PERCENTAGE, eqaul_size=True):
    file_list = rglob(data_root, PACKET_FILE_EXT)
    v2ray_file_list = [file_path for file_path in file_list if binary_classification(file_path) == 1]
    non_v2ray_file_list = [file_path for file_path in file_list if binary_classification(file_path) == 0]

    if eqaul_size:
        cut_off_count = min(len(v2ray_file_list), len(non_v2ray_file_list))
        v2ray_file_size = cut_off_count
        non_v2ray_file_size = cut_off_count
    else:
        v2ray_file_size = len(v2ray_file_list)
        non_v2ray_file_size = len(non_v2ray_file_list)

    v2ray_indexes = np.arange(len(v2ray_file_list))
    np.random.shuffle(v2ray_indexes)
    non_v2ray_indexes = np.arange(len(non_v2ray_file_list))
    np.random.shuffle(non_v2ray_indexes)

    training_file_list = [v2ray_file_list[index]
                          for index in v2ray_indexes[:math.ceil(v2ray_file_size * training_pct)]] + \
                         [non_v2ray_file_list[index]
                          for index in non_v2ray_indexes[:math.ceil(non_v2ray_file_size * training_pct)]]

    validation_file_list = [v2ray_file_list[index]
                            for index in v2ray_indexes[math.ceil(v2ray_file_size * training_pct): v2ray_file_size]] + \
                           [non_v2ray_file_list[index]
                            for index in non_v2ray_indexes[math.ceil(non_v2ray_file_size * training_pct): non_v2ray_file_size]]

    print("Statistics: ")
    print("Total V2ray traffic %d, Total non-V2ray traffic %d" % (len(v2ray_file_list), len(non_v2ray_file_list)))
    print("Output train traffic %d, Total validation traffic %d" % (len(training_file_list), len(validation_file_list)))

    return training_file_list, validation_file_list
# Generate training data and validation file list
train_file_list, val_file_list = generate_train_validation_packet_path_list(data_root=DATA_ROOT, eqaul_size=True)
````

Related issue/discussion:

https://github.com/v2fly/v2ray-core/issues/557
https://github.com/v2ray/discussion/issues/569

Ref:
https://datascience.stackexchange.com/questions/61858/oversampling-undersampling-only-train-set-only-or-both-train-and-validation-set

**Comment by rickyzhang82** [comment]
Both training set and validation set are isolated. There are no leaking. They are independent from each other.

As you can see the Python notebook written in last year, the trained model was loaded to test on unseen traffic, which are not from training set nor validation set.

**Comment by darsvador** [comment]
If you undersample the val/test set, it means that you assume that the ratio of real-world samples is also 1:1, which is unreasonable even though you have never used val/test set data for training.

**Comment by rickyzhang82** [comment]
I understood your points. But you overlook the fact that I could generate as many test data as I want.

In the testing step, I ran through the whole new unseen data set which are different from training and validation.

See the Python notebook 

```
Step 5: More Evaluation with unseen V2Ray traffic
```

**Comment by darsvador** [comment]
Thank you for your reply.
I have no problem with the notebook test set. But I think the ratio of positive and negative examples in the real world must be much lower than 1/4 which is the ratio in the notebook.
Let's back to the issue. https://github.com/v2fly/v2ray-core/issues/557

We have the dataset.
````
V2ray connections 17,998, none-V2ray TLS connections 136,981. 

17998/13698~=1/7
````
The ratio of positive and negative examples is about 1/7.
Even if the model predicts all examples to negative, the accuracy rate is as high as 88%.
The model predicts may be high recall rate and low precision. What about the model PR-curves?
Maybe it can better reflect the `precision` of the model.

**Comment by darsvador** [comment]
Another point I want to express is that if the ratio of positive and negative examples is not close to the real world during model testing, model prediction is of little practical significance. GFW obviously can‘t apply such a model to censorship. Nor can we explain how to improve our anti-censorship ability.

**Comment by rickyzhang82** [comment]
@DuckVador 

The ROC curve I plot in the V2Ray issue is on the evaluation data set, where both V2Ray (positive) and None-V2ray (negative) share the same size. I set aside 80% of data for training and 20% for evaluation. Both the positive and the negative in both training and evaluation has the equal size. That's where you complained "under sample" --  none-V2Ray traffic got thrown out.

The ROC curves shows true positive rate and false positive rate. You could derive the information from ROC curve for PR curves.

Note that

```
True Positive Rate = True Positive / Positive
False Positive Rate = False Positive / Negative
```

If all positive case were classified as negative, you could have a TPR = 0%. 

But I saw `0.99261029`% TPR, `0`% FPR when I ran the trained model to classify all data I have collected so far (Total V2ray traffic 17,998, Total non-V2ray traffic 136,981). To your questions, the precision is `100`%. 

You could argue that 80% of them is known to the classifier. It is cheating. But I lost the training file list and the evaluation file list. I could collect a complete new set of traffic data from a different V2Ray UUID, a server hosted by a different Cloud vendor and etc. But I don't think it would make a difference here as I did last year.

The real world traffic for sure is quite different from what I collected in a North American household. But the methodology is sound. 

GFW knows you are using V2Ray by a simple adaptive CNN classifier. You could be a sitting duck.

Good luck! Happy New Year.
